packages/api-contract/src/MetaRegistry.ts

```diff
@@ -0,0 +1,333 @@
+    return def === 'builtin';
+  }
+
+  private typeDefForBuiltin ({ id }: t.MetaType, typeIndex?: t.TypeIndex): Pick<t.TypeDef, never> {
+    if (!id) {
+      throw new Error('Invalid builtin type metadata found');
+    }
+
+    // builtin primitive type
+    if (typeof id === 'string') {
+      return {
+        info: t.TypeDefInfo.Plain,
+        type: id
+      };
+    }
+
+    // builtin tuple
+    if (Array.isArray(id)) {
+      const sub = id.map((tupleTypeIndex: number): t.TypeDef => this.typeDefFromMetaTypeAt(tupleTypeIndex));
+      return {
+        info: t.TypeDefInfo.Tuple,
+        type: encodeTuple(sub),
+        sub
+      };
+    }
+
+    // builtin array
+    const { 'array.type': vecTypeIndex, 'array.len': vecLength } = id as t.MetaTypeIdArray;
+    if (vecTypeIndex) {
+      assert(!vecLength || vecLength <= 256, 'MetaRegistry: Only support for [Type; <length>], where length <= 256');
+      assert(!typeIndex || vecTypeIndex !== typeIndex, `MetaRegistry: self-referencing registry type at index ${typeIndex}`);
+      const sub = this.typeDefFromMetaTypeAt(vecTypeIndex);
+
+      return {
+        ...(
+          vecLength
+            ? {
+              info: t.TypeDefInfo.VecFixed,
+              ext: { length: vecLength, type: '' },
+              type: `[${sub.name || sub.type};${vecLength}]`
+            }
+            : {
+              info: t.TypeDefInfo.Vec,
+              type: `Vec<${sub.name || sub.type}>`
+            }
+        ),
+        sub: this.typeDefFromMetaTypeAt(vecTypeIndex)
+      };
+    }
+
+    return {};
+  }
+
+  private isEnum ({ def }: t.MetaType): boolean {
+    const { 'enum.variants': enumVariants } = def as t.MetaTypeDefEnum;
+    return !!enumVariants;
+  }
+
+  private typeDefForEnum (def: t.MetaTypeDefEnum): t.TypeDef {
+    const sub = def['enum.variants'].map(
+      (variant): t.TypeDef => {
+        const { 'unit_variant.name': unitNameIndex } = variant as t.MetaTypeDefEnumVariantUnit;
+        if (unitNameIndex) {
+          return {
+            info: t.TypeDefInfo.Plain,
+            name: this.stringAt(unitNameIndex),
+            type: 'Null'
+          };
+        }
+
+        const { 'tuple_struct_variant.name': tupleStructVariantNameIndex } = variant as t.MetaTypeDefEnumVariantTupleStruct;
+        if (tupleStructVariantNameIndex) {
+          return this.typeDefForTupleStruct(variant as t.MetaTypeDefEnumVariantTupleStruct);
+        }
+
+        const { 'struct_variant.name': structVariantNameIndex } = variant as t.MetaTypeDefEnumVariantStruct;
+        if (structVariantNameIndex) {
+          return this.typeDefForStruct(variant as t.MetaTypeDefEnumVariantStruct);
+        }
+
+        return {
+          info: t.TypeDefInfo.Null,
+          type: 'Null'
+        };
+      }
+    );
+
+    return {
+      info: t.TypeDefInfo.Enum,
+      type: encodeEnum(sub),
+      sub
+    };
+  }
+
+  private isClikeEnum ({ def }: t.MetaType): boolean {
```

Comment:
maybe a typeguard here would be useful? `: def is t.MetaTypeDefClikeEnum` (if that's possible, because `def` is nested). So that we can avoid some `as ...`

Reply:
good idea but it looks like the destructuring does mess with it :(